### PR TITLE
Added unstructured io parsers

### PR DESCRIPTION
**Pull request by S1LV3RJ1NX**
- Cognita now supports the following file extensions: `".txt",
        ".eml",
        ".msg",
        ".xml",
        ".html",
        ".md",
        ".rst",
        ".json",
        ".rtf",
        ".jpeg",
        ".png",
        ".doc",
        ".docx",
        ".ppt",
        ".pptx",
        ".pdf",
        ".odt",
        ".epub",
        ".csv",
        ".tsv",
        ".xlsx"`
- Removed the other parsers to make files sleek
- Modified multimodal parser to directly support images: `".png", ".jpeg", ".jpg"`
- Minor modifications in parser pydantic in FE and BE.

### CODE REVIEW

File: backend/modules/parsers/unstructured_io.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+                    "strategy": "auto",
+                    "languages": ["eng", "hin"],
+                    "chunking_strategy": "by_title",
+                    "max_characters": self.max_chunk_size,
+                }
+
+                # Send POST request
+                response = self.session.post(
+                    settings.UNSTUCTURED_IO_URL.rstrip("/") + "/general/v0/general",
+                    headers=self.headers,
```
Comment (innoavator) [inline backend/modules/parsers/unstructured_io.py:79]: inline the headers value

File: docker-compose.yaml [hunk 2]
```diff
@@ -138,6 +169,7 @@ services:
             - TFY_API_KEY=${TFY_API_KEY}
             - TFY_HOST=${TFY_HOST}
             - DATABASE_URL=postgresql://${POSTGRES_USER}:${POSTGRES_PASSWORD}@cognita-db:5432/cognita-config
+            - UNSTUCTURED_IO_URL=${UNSTUCTURED_IO_URL}
```
Comment (chiragjn) [inline docker-compose.yaml:172]:
```suggestion
            - UNSTRUCTURED_IO_URL=${UNSTRUCTURED_IO_URL}
```

File: backend/Dockerfile [hunk 3]
```diff
@@ -40,6 +32,12 @@ ENV MODELS_CONFIG_PATH=${MODELS_CONFIG_PATH}
 ARG INFINITY_API_KEY
 ENV INFINITY_API_KEY=${INFINITY_API_KEY}
 
+ARG UNSTUCTURED_IO_URL
+ENV UNSTUCTURED_IO_URL=${UNSTUCTURED_IO_URL}
+
+ARG UNSTUCTURED_IO_API_KEY
+ENV UNSTUCTURED_IO_API_KEY=${UNSTUCTURED_IO_API_KEY}
```
Comment (chiragjn) [inline backend/Dockerfile:39]:
```suggestion
ARG UNSTRUCTURED_IO_URL
ENV UNSTRUCTURED_IO_URL=${UNSTRUCTURED_IO_URL}

ARG UNSTRUCTURED_IO_API_KEY
ENV UNSTRUCTURED_IO_API_KEY=${UNSTRUCTURED_IO_API_KEY}
```

File: frontend/src/screens/dashboard/docsqa/AddDataSourceToCollection.tsx [hunk 4]
```diff
@@ -14,12 +14,15 @@ import { MenuItem, Select } from '@mui/material'
...
   }
```
Comment (chiragjn) [inline frontend/src/screens/dashboard/docsqa/AddDataSourceToCollection.tsx:26]: Please let's not stuff these things in `additional_config`
Can we please refactor 

```
".pdf": {
    "class": "MultiModalParser",
    "kwargs": {
        "...": "..."
    }
```

File: docker-compose.yaml [hunk 5]
```diff
@@ -98,14 +98,42 @@ services:
         networks:
             - cognita-docker
 
+    cognita-parser:
```
Comment (chiragjn) [inline docker-compose.yaml:101]:
```suggestion
    unstructured-io-parsers:
```

File: compose.env [hunk 6]
```diff
@@ -19,6 +19,7 @@ MODELS_CONFIG_PATH="./models_config.yaml"
 METADATA_STORE_CONFIG='{"provider":"prisma"}'
 VECTOR_DB_CONFIG='{"provider":"qdrant","url":"http://qdrant-server:6333", "config": {"grpc_port": 6334, "prefer_grpc": false}}'
 COGNITA_BACKEND_PORT=8000
+UNSTUCTURED_IO_URL=http://cognita-parser:9500/
```
Comment (chiragjn) [inline compose.env:22]:
```suggestion
UNSTRUCTURED_IO_URL=http://unstructured-io-parsers:9500/
```

File: backend/settings.py [hunk 7]
```diff
@@ -57,3 +58,5 @@ def _validate_values(cls, values):
 
 
 settings = Settings()
+
+print(settings)
```
Comment (chiragjn) [inline backend/settings.py:62]: remove

File: backend/modules/parsers/unstructured_io.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+            backoff_factor=1,
+            status_forcelist=[429, 500, 502, 503, 504],
+            allowed_methods=["POST"],
+        )
+        self.adapter = HTTPAdapter(max_retries=self.retry_strategy)
+        self.session.mount("https://", self.adapter)
+        self.session.mount("http://", self.adapter)
+        self.headers = {
+            "accept": "application/json",
+        }
```
Comment (chiragjn) [inline backend/modules/parsers/unstructured_io.py:56]: All this retrying should be made a common utility across the repo

File: backend/modules/parsers/unstructured_io.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+        if settings.UNSTUCTURED_IO_API_KEY:
+            self.headers["unstructured-api-key"] = settings.UNSTUCTURED_IO_API_KEY
+
+    async def get_chunks(self, filepath: str, metadata: dict, *args, **kwargs):
+        """
+        Asynchronously extracts text from unstructured input and returns it in chunks.
+        """
+        final_texts = list()
```
Comment (chiragjn) [inline backend/modules/parsers/unstructured_io.py:64]:
```suggestion
        final_texts = []
```

File: backend/modules/parsers/unstructured_io.py [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+                }
+
+                # Send POST request
+                response = self.session.post(
+                    settings.UNSTUCTURED_IO_URL.rstrip("/") + "/general/v0/general",
+                    headers=self.headers,
+                    files=files,
+                    data=data,
+                )
+
```
Comment (chiragjn) [inline backend/modules/parsers/unstructured_io.py:83]:
```suggestion
response.raise_for_status()
```

File: backend/modules/parsers/unstructured_io.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+            parsed_data = response.json()
+            for payload in parsed_data:
+                text = payload["text"]
+                metadata = payload["metadata"]
+                final_texts.append(Document(page_content=text, metadata=metadata))
+            return final_texts
+        except Exception as e:
+            logger.exception(f"Final Exception: {e}")
+            return final_texts
```
Comment (chiragjn) [inline backend/modules/parsers/unstructured_io.py:92]: This is too aggressive error handling in my opinion, the caller should decide what they want to do this errors
This is general comment across all parsers that we maintain

File: backend/modules/parsers/unstructured_io.py (hunk 8 above)
Comment (S1LV3RJ1NX) [reply backend/modules/parsers/unstructured_io.py:56]: will take this up as a separate PR

File: backend/modules/parsers/unstructured_io.py [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+        Asynchronously extracts text from unstructured input and returns it in chunks.
+        """
+        final_texts = list()
+        try:
+            with open(filepath, "rb") as f:
+                # Define files payload
+                files = {"files": f}
+                data = {
+                    "strategy": "auto",
+                    "languages": ["eng", "hin"],
```
[the remark this answers is not in the document]
Comment (S1LV3RJ1NX) [reply backend/modules/parsers/unstructured_io.py:71]: this I'll take it up in parser config PR.

**Review by innoavator — APPROVED** [review]

File: backend/modules/parsers/multimodalparser.py [hunk 13]
```diff
@@ -153,38 +144,57 @@ async def get_chunks(
...
+                not filepath.endswith(".pdf")
+                and not filepath.endswith(".png")
+                and not filepath.endswith(".jpeg")
+                and not filepath.endswith(".jpg")
+            ):
                 logger.error(
-                    "Invalid file extension. MultiModalParser only supports PDF files."
+                    "Invalid file extension. MultiModalParser only supports PDF, PNG, JPEG, JPG files."
                 )
                 return []
```
Comment (chiragjn) [inline backend/modules/parsers/multimodalparser.py:156]: Again, in my opinion, parsers should raise exceptions and callers should decide what they want to do with the error

File: backend/modules/parsers/multimodalparser.py [hunk 14]
```diff
@@ -153,38 +144,57 @@ async def get_chunks(
...
-                    _, buffer = cv2.imencode(".png", img)
-                    image_base64 = base64.b64encode(buffer).decode("utf-8")
 
-                    pages[page_number] = image_base64
-                except Exception as e:
-                    logger.exception(f"Error in page: {page_number} - {e}")
-                    continue
+            if filepath.endswith(".pdf"):
+                # Open the PDF file using pdfplumber
+                doc = fitz.open(filepath)
```
Comment (chiragjn) [inline backend/modules/parsers/multimodalparser.py:163]: How is ftiz requirement managed? Is it in core requirements?

File: backend/modules/parsers/multimodalparser.py [hunk 15]
```diff
@@ -153,38 +144,57 @@ async def get_chunks(
...
+                pages = dict()
```
Comment (chiragjn) [inline backend/modules/parsers/multimodalparser.py:164]: nit
```suggestion
                pages = {}
```

**Review by chiragjn — APPROVED** [review]
Keeping comments which will be addressed later as unresolved

**Review by chiragjn — APPROVED** [review]

**Pull request merged**